### PR TITLE
Create DataFixer system for mods

### PR DESCRIPTION
**Pull request by itsmeow**
Draft PR for obvious reasons, it still doesn't work. I would just like some early-ish feedback so I can make this better.

### CODE REVIEW

**Comment by williewillus** [comment]
you'll have to explain what you can of the vanilla system to us, since I doubt anyone understands it more than you do ;p

**Comment by itsmeow** [comment]
Right. I think Gamebuster explained it pretty well [here](https://github.com/MinecraftForge/MinecraftForge/issues/5695#issuecomment-483465503)

**Comment by itsmeow** [comment]
To address concerns there:
> ## The current (known) problems I have found are as follows:
> 
>     1. Minecraft currently only supports one data version. We would need some sort of system that allows multiple DataFixerUppers, and each DataFixerUpper would need its own version. We would also need to need to save the current fixer version for each mod when we save data.
I have not yet made a system that saves forge versions to the world on load, but it is set up to load it already from a compound called "ForgeVersionData" which has integers with the key being a modid and value the data version that is saved.
This is solved via the static `builders` map in `DataFixers`, which also has `init(String modid, int version)` which constructs and adds your fixer, as well as returns it (it should be stored locally by whatever mod uses it, they should NOT be using `getBuilders().get(modid)` as Forge does)
>     2. If the above is implemented, what should we do if two or more DataFixerUppers try to fix the same set of data? Personally I'd crash since there could irreversible data corruption.
Right now I haven't put logic for this, I'm not even sure if it's working. But the way vanilla works is it has all of its initial datafixing logic in `V099`, which dictates what happens when a basic type is not present (key for modded data etc) and what happens (it's kept/it's thrown out),  so all mods should go BEFORE vanilla. Maybe two stages of datafixing that mods can have separate fixers for?
>     3. Some data fixing is done outside of saving/loading worlds (I think I remember seeing options getting datafixed). So there is a possibility that some datafixing is being done before mod loading. I'm not 100% certain on this though.
The only datafixing outside of worlds is options and advancements. Currently I have ignored both as they are mostly useless to modders.

**Comment by itsmeow** [comment]
In order to allow changes after the mod has been initialized, I've made `ModFixes` extend the builder class and allow building on the instance, but it also stores a fixer instance which gets set to null every time a schema or fixer is added to the builder. It is rebuilt when the fixer is next required if the stored instance is null.

**Comment by itsmeow** [comment]
As for an example, in mod construction:
```java
        fixes = ModFixes.init(Ref.MOD_ID, 100);
        fixes.addSchema(0, V0099::new);
        Schema schema = fixes.addSchema(99, BASchema::new);
        fixes.addFixer(new AddNewChoices(schema, "Add BA+ tileentity types", TypeReferences.BLOCK_ENTITY));
        fixes.addFixer(new AddNewChoices(schema, "Add BA+ entity types", TypeReferences.ENTITY_TYPE));
        fixes.addSchema(1, Schema::new);
```
BASchema overrides:
```java
    @Override
    public Map<String, Supplier<TypeTemplate>> registerEntities(Schema schema) {
        Map<String, Supplier<TypeTemplate>> map = Maps.newHashMap();
        for(EntityContainer container : ModEntities.entityList) {
            schema.registerSimple(map, Ref.MOD_ID + ":" + container.entityName);
        }
        return map;
    }

    @Override
    public Map<String, Supplier<TypeTemplate>> registerBlockEntities(Schema schema) {
        Map<String, Supplier<TypeTemplate>> map = Maps.newHashMap();
        schema.registerSimple(map, Ref.MOD_ID + ":head");
        schema.registerSimple(map, Ref.MOD_ID + ":trilliumtileentity");
        schema.registerSimple(map, Ref.MOD_ID + ":handoffatetileentity");
        return map;
    }
```
Although I haven't got this fully working yet has it creates some weird low-level errors with the fixers. Specifically, LeavesFix:
```
java.lang.ClassCastException: java.util.ArrayList cannot be cast to com.mojang.datafixers.Typed
	at net.minecraft.util.datafix.fixes.LeavesFix$Section.write(LeavesFix.java:317) ~[main/:?]
```

**Comment by itsmeow** [comment]
Figured out what's causing the issue. So, V099 is the only schema that overrides `registerTypes`, which is responsible for determining what each TypeReference (equivalent to 1.12 datafix types, such as CHUNK or ITEM_STACK or BLOCK_ENTITY) does by default. This is also what decides what happens if a key is nil, and whether the type is recursive (it should check for itself in itself). So the FIRST schema registered on a fixer needs to override registerTypes, registerEntities, and registerBlockEntities, register each type it will be using to `DSL::remainder`, and return a new hash map of <String, Supplier<TypeTemplate>> for the registerEntity methods.
`DSL::remainder` is also known as NIL_SAVE. Basically, if key is nil, save the data. Vanilla seems to like throwing out data it doesn't find. This is why I propose mods go before vanilla. We might need to modify vanilla to keep data it doesn't know by default, although I can't be sure what the implications of this are.

**Comment by itsmeow** [comment]
Code responsible for throwing out data that needs to be accounted for is in TaggedChoice (this is of the DataFixerUpper library) in com.mojang.datafixers.types.templates
Starting at line 195:
```java
                        if (DataFixerUpper.ERRORS_ARE_FATAL) {
                            throw new IllegalArgumentException("Unsupported key: " + key.get() + " in " + this);
                        } else {
                            LOGGER.warn("Unsupported key: {} in {}", key.get(), this);
                            return Pair.of(input, Optional.empty());
                        }
```
`Optional.empty()` is functionally equivalent to throwing out data as it is meant to perform a given operation on the data and return it back with that data here. Either we're doing something wrong, MC is doing something wrong, or the lib is doing something wrong. I'm not quite sure.

**Comment by itsmeow** [comment]
The above is also responsible for log spam:

> [mojang/TaggedChoice]: Unsupported key: betteranimalsplus:deer in TaggedChoiceType[id, minecraft:spawner_minecart -> ((Tag["DisplayTile", NamedType["block_name", (Int | NamespacedString)]] | NilDrop), MuType[V1125_4]), 
minecraft:giant -> ((Tag["ArmorItems", List[MuType[V1125_3]]] | NilDrop), ((Tag["HandItems", List[MuType[V1125_3]]] | NilDrop), NilSave)), 
minecraft:tnt -> NilSave,

**Comment by itsmeow** [comment]
I've created a similar wrapper for DataFixesManager that acts like ModFixes but for once static instance (vanilla), so that builder entries can be added after static initialization and then be rebuilt. It works pretty well (not sure on how performance/memory heavy it is) and I am current working on a test mod to register entity and tileentity types. This will hopefully be an automatic registration of datafixers so that MC does not throw out the data. I'm still not sure if ModFixes is going to be something that works. The main goal here is to get vanilla to not throw out data, which is what I made #5695 to address.

**Comment by itsmeow** [comment]
On new examples that work.
I've got Entity Types to register to vanilla, blockentities aren't working just yet. Anyways, registering to vanilla is this simple (using same BASchema class as before):
```java
        DataFixesManager vanillaFixes = DataFixesManager.getBuilder();
        Schema vSchema = vanillaFixes.addSchema(DataFixesManager.FIXER_VERSION - 1, BATestSchema::new);
        vanillaFixes.addFixer(new AddNewChoices(vSchema, "Add BA+ tileentity types", TypeReferences.BLOCK_ENTITY));
        vanillaFixes.addFixer(new AddNewChoices(vSchema, "Add BA+ entity types", TypeReferences.ENTITY_TYPE));
```
This should be movable to forge in some form so mods will automatically keep data.

File: src/main/java/net/minecraftforge/fml/fixers/ModFixes.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        this.dataVersion = dataVersion;
+        this.mod = modid;
+    }
+
+    public String getModId()
+    {
+        return mod;
+    }
+
+    public DataFixer getBuiltFixer()
```
Comment (Gamebuster19901) [inline src/main/java/net/minecraftforge/fml/fixers/ModFixes.java:138]: Generally you don't want your datafixer to have behavior that changes. For each fixer version, it should fix data the exact same way every single time. I'm not really a fan of allowing mod fixers to be rebuilt.

**Review by Gamebuster19901 — CHANGES_REQUESTED** [review]
I don't think mutable datafixers are a good idea. I can't think of a reason to add fixers/schemas after the fixer is built. Fixers should behave the exact same way every single time for a specific fixer version.

File: src/main/java/net/minecraftforge/fml/fixers/ModFixes.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    public static Map<DataFixer, Pair<Integer, Integer>> getFixVersions(NBTTagCompound nbttagcompound, boolean useLatestOtherwiseN1)
+    {
+        return getFixVersions(nbttagcompound, (useLatestOtherwiseN1 ? DataFixesManager.FIXER_VERSION : -1), (useLatestOtherwiseN1 ? -2492 : -1));
+    }
+
+    public static Map<DataFixer, Pair<Integer, Integer>> getFixVersionsNoVanilla(NBTTagCompound nbttagcompound, boolean useLatestOtherwiseN1)
+    {
+        return getFixVersionsNoVanilla(nbttagcompound, (useLatestOtherwiseN1 ? -2492 : -1));
```
Comment (Gamebuster19901) [inline src/main/java/net/minecraftforge/fml/fixers/ModFixes.java]: What is -2492?

Comment (itsmeow) [reply src/main/java/net/minecraftforge/fml/fixers/ModFixes.java]: Arbitrary number, it's passed to the above. Kind of a bad solution. I will fix this.

**Comment by itsmeow** [comment]
Mutable fixers is not the idea here. The issue is MC likes to check fixers during static initialization: see EntityType `register()` so I need the fixer to rebuild at least once before world load. Perhaps all fixers are rebuilt after construction and stay that way, and anything added afterwards would be ignored/throw an error?

**Comment by itsmeow** [comment]
The reason fixers need to be built during construction: Registration of entities needs to take place after the fixer is built, otherwise entitytype and tileentitytype fixers cannot be found and will log an error, even if there is none.

**Comment by itsmeow** [comment]
I will be removing the ability to rebuild mod fixers, but vanilla fixers have to be rebuilt by default.

**Comment by Gamebuster19901** [comment]
I didn't even realize that. I think disallowing mod fixers to be rebuilt will work.

**Comment by itsmeow** [comment]
Alright, @Gamebuster19901, I've made the changes as well as added saving versions for mod fixers. I will test this soon.

**Comment by itsmeow** [comment]
Alright, this next one changes a lot of what a test mod would look like. This is what I am using (trimmed other construction stuff):
```java
    public BetterAnimalsPlusMod() {
        //trim
        for(HeadTypes type : HeadTypes.values()) {
            String name =  Ref.MOD_ID + ":" + type.name;
            registerFlattenedNSWEUD(name);
            ItemStackDataFlattening.registerSimpleFlattening(name, name);
        }
        registerFlattenedNSWE(Ref.MOD_ID + ":trillium");
        registerFlattenedNSWE(Ref.MOD_ID + ":handoffate");
        
        ForgeTypesSchema.registerSimpleTileEntity(Ref.MOD_ID + ":head");
        ForgeTypesSchema.registerSimpleTileEntity(Ref.MOD_ID + ":trilliumtileentity");
        ForgeTypesSchema.registerSimpleTileEntity(Ref.MOD_ID + ":handoffatetileentity");
        for(EntityContainer container : ModEntities.entityList) {
            ForgeTypesSchema.registerEquipmentEntity(Ref.MOD_ID + ":" + container.entityName);
        }
        ForgeTypesSchema.registerSimpleEntity("betteranimalsplus:badgerdirt");
        ForgeTypesSchema.registerSimpleEntity("betteranimalsplus:tarantulahair");

        //trim
    }
    
    private static void registerFlattenedNSWE(String name) {
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'north'}}", "{Name:'" + name + "',Properties:{facing:'north'}}");
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'south'}}", "{Name:'" + name + "',Properties:{facing:'south'}}");
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'west'}}", "{Name:'" + name + "',Properties:{facing:'west'}}");
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'east'}}", "{Name:'" + name + "',Properties:{facing:'east'}}");
    }
    
    private static void registerFlattenedNSWEUD(String name) {
        registerFlattenedNSWE(name);
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'up'}}", "{Name:'" + name + "',Properties:{facing:'up'}}");
        BlockStateFlatteningMap.addEntry("{Name:'" + name + "',Properties:{facing:'down'}}", "{Name:'" + name + "',Properties:{facing:'down'}}");
    }
```
The advantage of this is now users should never have to actually add a fixer, Forge adds one which can be registered to. This prevents ID conflicts.

**Comment by itsmeow** [comment]
I'm a little stumped here, I need someone else to take a look. I load a world from my mod with fixer version 1343, and my blocks, itemstacks, TEs, and entities are still being thrown out. I can't determine a source and there's no log output. Although I'm still getting some really bad log spam from the LeavesFix error, and I am unable to determine the source of it:
```
[Server thread/ERROR] [mojang/DataFixerUpper]: Something went wrong upgrading!
java.lang.ClassCastException: java.util.ArrayList cannot be cast to com.mojang.datafixers.Typed
	at net.minecraft.util.datafix.fixes.LeavesFix$Section.write(LeavesFix.java:317) ~[main/:?]
	at net.minecraft.util.datafix.fixes.LeavesFix.lambda$8(LeavesFix.java:129) ~[main/:?]
	at com.mojang.datafixers.Typed.lambda$updateTyped$6(Typed.java:125) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.lambda$updateCap$9(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.ListTraversal.lambda$wander$0(ListTraversal.java:19) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateCap(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:124) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:119) ~[datafixerupper-1.0.19.jar:?]
	at net.minecraft.util.datafix.fixes.LeavesFix.lambda$3(LeavesFix.java:128) ~[main/:?]
	at com.mojang.datafixers.Typed.lambda$updateTyped$6(Typed.java:125) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.lambda$updateCap$9(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Either$Instance.lambda$traverse$6(Either.java:227) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Either$Left.map(Either.java:38) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Either$Instance.traverse(Either.java:225) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateCap(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:124) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:119) ~[datafixerupper-1.0.19.jar:?]
	at net.minecraft.util.datafix.fixes.LeavesFix.lambda$2(LeavesFix.java:65) ~[main/:?]
	at com.mojang.datafixers.Typed.lambda$updateTyped$6(Typed.java:125) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.lambda$updateCap$9(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.util.Pair$Instance.traverse(Pair.java:83) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.profunctors.TraversalP$1.lambda$wander$0(TraversalP.java:50) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.optics.Traversal$Instance$1.lambda$wander$0(Traversal.java:34) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateCap(Typed.java:162) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:124) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.Typed.updateTyped(Typed.java:119) ~[datafixerupper-1.0.19.jar:?]
	at net.minecraft.util.datafix.fixes.LeavesFix.lambda$1(LeavesFix.java:63) ~[main/:?]
	at com.mojang.datafixers.DataFix.lambda$null$4(DataFix.java:75) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.functions.Comp.lambda$null$5(Comp.java:69) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.functions.Comp.lambda$null$5(Comp.java:69) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.functions.Comp.lambda$null$5(Comp.java:69) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.functions.Comp.lambda$null$5(Comp.java:69) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.functions.Comp.lambda$null$5(Comp.java:69) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.types.Type.capWrite(Type.java:154) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.types.Type.lambda$null$7(Type.java:145) ~[datafixerupper-1.0.19.jar:?]
	at java.util.Optional.map(Unknown Source) ~[?:1.8.0_201]
	at com.mojang.datafixers.types.Type.lambda$readAndWrite$8(Type.java:144) ~[datafixerupper-1.0.19.jar:?]
	at java.util.Optional.flatMap(Unknown Source) ~[?:1.8.0_201]
	at com.mojang.datafixers.types.Type.readAndWrite(Type.java:143) ~[datafixerupper-1.0.19.jar:?]
	at com.mojang.datafixers.DataFixerUpper.update(DataFixerUpper.java:83) [datafixerupper-1.0.19.jar:?]
	at net.minecraft.nbt.NBTUtil.updateSingle(NBTUtil.java:270) [?:?]
	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_212146_a(AnvilChunkLoader.java:106) [?:?]
	at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_208030_a(AnvilChunkLoader.java:81) [?:?]
	at net.minecraft.world.chunk.storage.AnvilChunkLoader.loadChunk(AnvilChunkLoader.java:142) [?:?]
	at net.minecraft.world.gen.ChunkProviderServer.provideChunk(ChunkProviderServer.java:97) [?:?]
	at net.minecraft.server.management.PlayerChunkMapEntry.<init>(PlayerChunkMapEntry.java:43) [?:?]
	at net.minecraft.server.management.PlayerChunkMap.getOrCreateEntry(PlayerChunkMap.java:178) [?:?]
	at net.minecraft.server.management.PlayerChunkMap.addPlayer(PlayerChunkMap.java:211) [?:?]
	at net.minecraft.server.management.PlayerList.preparePlayer(PlayerList.java:250) [?:?]
	at net.minecraft.server.management.PlayerList.playerLoggedIn(PlayerList.java:308) [?:?]
	at net.minecraft.server.management.PlayerList.initializeConnectionToPlayer(PlayerList.java:151) [?:?]
	at net.minecraft.network.NetHandlerLoginServer.tryAcceptPlayer(NetHandlerLoginServer.java:113) [?:?]
	at net.minecraft.network.NetHandlerLoginServer.tick(NetHandlerLoginServer.java:64) [?:?]
	at net.minecraft.network.NetworkManager.tick(NetworkManager.java:235) [?:?]
	at net.minecraft.network.NetworkSystem.tick(NetworkSystem.java:134) [?:?]
	at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:765) [?:?]
	at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:665) [?:?]
	at net.minecraft.server.integrated.IntegratedServer.tick(IntegratedServer.java:115) [?:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:561) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_201]
```

**Comment by Gamebuster19901** [comment]
@itsmeow maybe LeavesFix.Section#isSkippable() is returning false when it should be true?

LeavesFix.Section.storage may be non null when it should be null.

**Comment by Gamebuster19901** [comment]
I have a question though.

Would it be possible to add a com.mojang.Datafix that implements com.mojang.datafixers.DataFixer?

That hybrid fixer would contain all of the mod fixes.

then we can then add that hybrid fixer to DataFixesManager.addFixers().

Would this work?

**Pull request closed without merging**